### DISCUSSION TITLE
Integrating appium with code magic for native android.

### DISCUSSION THREAD

**Discussion by vivekgabadiya** [codemagic-ci-cd/codemagic-docs#1894] [Q&A]
### If you are looking for help, please confirm the following...

- [X] I confirm I have searched the [Docs](https://docs.codemagic.io), [Codemagic Sample Projects](https://github.com/codemagic-ci-cd/codemagic-sample-projects), and GitHub [Discussions](https://github.com/orgs/codemagic-ci-cd/discussions).
- [X] I confirm the software versions on my local computer match the ones in CI environment (e.g. Xcode version, Gradle version). See [here](https://docs.codemagic.io/troubleshooting/common-android-issues/#builds-work-locally-but-fail-on-codemagic)
- [X] I confirm I am able to do a fresh clone of the repository and build the project on my local computer.

### Which mobile framework are you using?

Android (Kotlin)

### Steps to reproduce

Sample code is available for running appium with code magic using react native framework. I am looking for steps to implement the same for my native android framework.

### Expected results

Appium test script should be running before publishing the app in market also we can stop the publishing process if the test result is failed.

### Actual results

No results

**Comment by himesh-cm** [comment]
Hi @vivekgabadiya, Please check the guide [here](https://docs.codemagic.io/yaml-testing/testing/#react-native-integration-tests-using-appium-and-emulator) for Appium and RN. Let us know if you hit any roadblocks.

**Reply by vivekgabadiya** [reply]
Hello Himesh, we are looking for guide about Android native and Appium but the link you suggested is for React native and Appium.

**Reply by himesh-cm** [reply]
Hi @vivekgabadiya, I noticed that there isn't a guide in the Codemagic documentation for using Native Android and Appium. However, I wanted to let you know that we have Appium pre-installed on our builder machines, which means that you can add any CLI commands you use for local testing as a script in your Codemagic workflows. I would suggest giving it a try and seeing how it goes. If you encounter any problems, please don't hesitate to share your build ID with us so we can assist you better.

**Reply by vivekgabadiya** [reply]
Hello @himesh-cm, we haven't updated yaml file to include script for adding appium. Please help us to update the script, our build ID is 64b7c27b5ec8bf2265fc60bb.

**Reply by himesh-cm** [reply]
Hi @vivekgabadiya, I am unaware of your project structure, testing scope, architecture, etc. Ideally, the command you use to run the tests locally is the same one you can use in the `codemagic.yaml` file.

**Comment by himesh-cm** [comment]
I'm tentatively closing the discussion since we haven't heard back from you. You can open the discussion again by replying to this message or opening a new discussion.

**Reply by vivekgabadiya** [reply]
Hello Himesh, Sorry for delay in reply. Before running our appium test cases we were trying to run simple UI test and we found that android simulator is not available in the builder machine. We also tried changing our builder machine to Mac Pro but it is giving error message "The mac_pro instance type does not support the selected Xcode version. Please consider using mac_mini_m1 or mac_mini_m2 instead."and linux machine is for paid users only. 

Please suggest how to proceed.

**Reply by himesh-cm** [reply]
@vivekgabadiya, What `Xcode` version are you using with `mac_pro` machines?

**Reply by himesh-cm** [reply]
If you're using Apple silicon, it's important to note that the Apple Virtualization API doesn't support nested virtualization, which is required for Android emulators. To keep using the free M1 minutes and run tests on Android emulators in macOS M1 workflows, you can use third-party service integrations like emulator.wtf, Katalon, Firebase, or AWS Device Farm.

**Reply by vivekgabadiya** [reply]
@himesh-cm I have tried Xcode version 14.2 and 12.5 and it was saying the selected instance type is not available with the current billing plan. So as per your comment we need paid account to try android emulators using code magic otherwise we have to use third-party Services, is that correct?

**Reply by himesh-cm** [reply]
Yes, you are correct. Mac Pro machines can only be used if you have billing enabled. Only M1 instances are available in personal teams free tier.

**Reply by vivekgabadiya** [reply]
@himesh-cm Thanks for the updates. We have configured firebase test lab for UI testing and using code magic for unit testing. Can you please help us with appium integration steps. We are currently writing appium test scripts using python and running it successfully using browser stack. We are looking forward to move it towards code magic.

**Reply by himesh-cm** [reply]
Hi @vivekgabadiya, Kindly share the steps you run `locally / browser stack`.

**Reply by vivekgabadiya** [reply]
Hi @himesh-cm, with the help of code shown below we are able to  invoke app in an iOS device on the souce_lab website. We can run this file by opening it in a pycharm editor and by running the main.py  file.  

**main.py file code**
from appium import webdriver
from appium.webdriver.common.appiumby import AppiumBy

class AppiumTest:
    def __init__(self):
        self.caps = {
            'platformName': 'iOS',
            'appium:app': 'storage:filename=TTMIOS.ipa',
            'appium:deviceName': 'iPhone 12',
            'appium:deviceOrientation': 'portrait',
            'appium:platformVersion': '16.4.1',
            'appium:automationName': 'XCUITest',
            'sauce:options': {
                'username': 'user',
                'accessKey': 'ksy',
                'build': 'appium-build',
                'name': '<your test name>'
            }
        }

    def run_test(self):
        # Set the Appium server URL
        url = 'https://ondemand.us-west-1.saucelabs.com:443/wd/hub'
        driver = webdriver.Remote(url, self.caps)

        # Your test script goes here
        driver.find_element(AppiumBy.ID("com.xyz.masterapp:id/EditText_username")).send_keys("liidartestdata@photogauge.com")

        # Quit the driver after the test is done
        driver.quit()


# Instantiate the class and run the test
if __name__ == "__main__":
    appium_test = AppiumTest()
    appium_test.run_test()

**Reply by himesh-cm** [reply]
Did you try running `main.py` file in your workflow?
Please share a build id.

**Reply by vivekgabadiya** [reply]
Hi @himesh-cm We tried running main.py file it is working fine. One more question to ask i.e. in case of instrumentation testing using  firebase test labs either we get result as pass or fail our workflow is proceeding for app publishing, how we can stop going for publishing in case our test result is failed. Thanks

**Reply by himesh-cm** [reply]
Hi, @vivekgabadiya It's wonderful to hear that you were able to make it work successfully. You can create a simple `if / else` bash script following these steps for conditional publishing.

Please change the script, as per your requirements.

```  
  - name: Test finished successfully
    script: touch ~/SUCCESS
  - name: publishing build
    scripts: | 
        if [  -f ~/SUCCESS ] ; then
           # publish the build
        else
           exit
        fi
 ```

**Reply by vivekgabadiya** [reply]
@himesh-cm Can you please provide some idea what will be script in case of firebase test lab result pass and failed.

**Reply by himesh-cm** [reply]
You can check out the docs [here](https://docs.codemagic.io/yaml-testing/firebase-test-lab/) for the firebase test lab.

**Comment by himesh-cm** [comment]
I'm tentatively closing the discussion since we haven't heard back from you. You can open the discussion again by replying to this message or opening a new discussion.
